### DISCUSSION TITLE
tagged_iterator autowire constructor

### DISCUSSION THREAD

**Discussion by benjaminfunk** [symfony/symfony#48720] [Q&A]
I have no idea whether I configured it correctly or whether it might even be an bug.

```yaml
    _instanceof:
        App\Template\TemplateInterface:
            autowire: false
            tags: ['app.template']

    App\Provider\TemplateProvider:
        arguments:
            - !tagged_iterator app.template
```

```php
class TemplateProvider
{
	private iterable $providers;

	public function __construct(iterable $providers)
	{
		$this->providers = $providers;
	}

	public function create(string $type, ProductInterface $product, CustomizerInterface $customizer): TemplateInterface
	{
		foreach ($this->providers as $provider)
		{
			if ($type === $provider::getType())
			{
				return new $provider($product, $customizer);
			}
		}

		throw new InvalidArgumentException('Unknown channel given');
	}
}
```

```php
abstract class Template implements TemplateInterface
{
	public static function getType(): string
	{
		return (new ReflectionClass(get_called_class()))->getShortName();
	}
}
```

```php
class WJ55 extends Template
{
	public function __construct(ProductInterface $product, BE98 $data)
	{
		$this->product = $product;
		$this->data = $data;
	}
}
```

```php
public function action(CustomizerProvider     $customizerprovider,
			ProductProvider        $productProvider,
			TemplateProvider       $templateProvider): Response
	{
		$customizer = $customizerprovider->create('BE98');
		$product = $productProvider->create('Canvas');

		$templateProvider->create('WJ55', $product, $customizer);

		return new Response('');
	}
```

This is my structure so that i can dynamically create the right class with a string. The provider is like a factory pattern but as a service instead of a static function.

The provider for "customizer" and "product" work exactly the same and they do the job without problems. However, the "template" has the difference that it has a constructor and thats why my app throw an exception. Thats why i disabled the autowire so that I can pass the parameters myself.

As soon as i iterate the "$this->provider" array in the provider the app throw an exception. The constructur should be executes as this point?

![Screenshot 2022-12-20 143940](https://user-images.githubusercontent.com/856290/208680629-0480d04f-5bf3-42c9-884f-8aca5c6a7d8f.jpg)

Any idea how i can fix this? Or is it maybe a bug, that the iteration call the constructor to soon?

The problem seems to be with the "tagged_iterator". If i change the tagged_iterator to a normal list it works...

```yaml
    App\Provider\TemplateProvider:
        arguments:
            - App\Template\WJ55
            - App\Template\GG85
            - App\Template\ER99
```

**Comment by ahundiak** [comment]
The problem is that your templates are not actually services but you add them to a service locator anyways.  Your last approach works because you are just providing a list of class names.  When you provide an array of services then the line `foreach ($this->providers as $provider)` tries to instantiate the actual service which of course fails because it has no idea what the dependencies are.

Personally I would just stick with your last approach and keep the list updated as additional templates are added.  As an alternative you could create a factory for each template and pass a tagged list of factories.  Doing so would allow injecting additional dependencies to the factory if you ever need to.  

Or you might get a little bit fancy and use a compiler pass to find all your tagged template non-services, inject an array of the template class names into your template provider and then remove the individual template non-services so they never get injected.  Bit convoluted but it should work.
